src/OmniSharp.MSBuild/ProjectFile/ProjectFileInfo.ProjectData.cs

```diff
@@ -286,41 +286,30 @@ public static ProjectData Create(MSB.Execution.ProjectInstance projectInstance,
 
                 var projectReferences = ImmutableArray.CreateBuilder<string>();
                 var projectReferenceAliases = ImmutableDictionary.CreateBuilder<string, string>();
-                var projectReferencesAdded = new HashSet<string>();
-                foreach (var projectReferenceItem in projectInstance.GetItems(ItemNames.ProjectReference))
-                {
-                    var fullPath = projectReferenceItem.GetMetadataValue(MetadataNames.FullPath);
-
-                    if (IsCSharpProject(fullPath) && projectReferencesAdded.Add(fullPath))
-                    {
-                        projectReferences.Add(fullPath);
-
-                        var aliases = projectReferenceItem.GetMetadataValue(MetadataNames.Aliases);
-                        if (!string.IsNullOrEmpty(aliases))
-                        {
-                            projectReferenceAliases[fullPath] = aliases;
-                        }
-                    }
-                }
 
                 var references = ImmutableArray.CreateBuilder<string>();
                 var referenceAliases = ImmutableDictionary.CreateBuilder<string, string>();
                 foreach (var referencePathItem in projectInstance.GetItems(ItemNames.ReferencePath))
                 {
                     var referenceSourceTarget = referencePathItem.GetMetadataValue(MetadataNames.ReferenceSourceTarget);
+                    var aliases = referencePathItem.GetMetadataValue(MetadataNames.Aliases);
 
+                    // If this reference came from a project reference, count it as such. We never want to directly look
+                    // at the ProjectReference items in the project, as those don't always create project references
+                    // if things like OutputItemType or ReferenceOutputAssembly are set. It's also possible that other
+                    // MSBuild logic is adding or removing properties too.
                     if (StringComparer.OrdinalIgnoreCase.Equals(referenceSourceTarget, ItemNames.ProjectReference))
                     {
-                        // If the reference was sourced from a project reference, we have two choices:
-                        //
-                        //   1. If the reference is a C# project reference, we shouldn't add it because it'll just duplicate
-                        //      the project reference.
-                        //   2. If the reference is *not* a C# project reference, we should keep this reference because the
-                        //      project reference was already removed.
-
-                        var originalItemSpec = referencePathItem.GetMetadataValue(MetadataNames.OriginalItemSpec);
-                        if (originalItemSpec.EndsWith(".csproj", StringComparison.OrdinalIgnoreCase))
+                        var sourceProjectFile = referencePathItem.GetMetadataValue(MetadataNames.MSBuildSourceProjectFile);
```

Comment:
Still confirming with the MSBuild team precise which property should be used here.

Reply:
* `OriginalItemSpec`: available on all RAR outputs. The `Include` value of the RAR input that caused that output. (This is actually automatically set in the engine whenever populating an item from another item)
* `ProjectReferenceOriginalItemSpec`: available on RAR outputs that were resolved from `@(ProjectReference)` items. The `OriginalItemSpec` value of the `Reference` that `ResolveProjectReferences` produced from the `ProjectReference` item--so the `ProjectReference`'s `Include`.
* `OriginalProjectReferenceItemSpec: ¯\\(°_o)/¯
  > Copy OriginalItemSpec to OriginalProjectReferenceItemSpec, so that when ResolveAssemblyReferences
  > takes these items and resolves them to ReferencePath, we can still recover the _real_ OriginalItemSpec
  > for the unresolved reference items.
  
  Doesn't appear to be used anywhere. My advice: avoid.
* `MSBuildSourceProjectFile`: available on all items that are outputs of the `<MSBuild` task. Points to the project that created the item. For RAR outputs, inherited from the Reference generated by `<MSBuild` in `ResolveProjectReferences`


So the thing that most matches your use case is `ProjectReferenceOriginalItemSpec`. It has a weakness though: it's a relative path, not fully resolved. As long as you deal with that I'd go with it. `MSBuildSourceProjectFile` should always be an absolute path which is nice but it could get confused with custom reference-resolution schemes.

Reply:
@rainersigwald What we noticed is that the fully qualified `MSBuildSourceProjectFile` prepends the `/private/` folder name on MacOS for the shadowcopied unit tests which causes them to be different that the paths we qualifyw with `Path.GetFullPath`. I'll take your advice and update to use `ProjectReferenceOriginalItemSpec`.